marlowe-symbolic/src/Language/Marlowe/Analysis/FSSemantics.hs

```diff
@@ -1,1126 +1,415 @@
+countWhens (Pay uv uw ux uy c) = countWhens c
+countWhens (If uz c c2)        = max (countWhens c) (countWhens c2)
+countWhens (When cl t c)       = 1 + max (countWhensCaseList cl) (countWhens c)
+countWhens (Let va vb c)       = countWhens c
+
+countWhensCaseList :: [Case Contract] -> Integer
+countWhensCaseList (Case uu c : tail) = max (countWhens c) (countWhensCaseList tail)
+countWhensCaseList []                 = 0
+
+wrapper :: Contract -> [(SInteger, SInteger, SInteger, SInteger)] -> Maybe State
+        -> Symbolic SBool
+wrapper c st maybeState = do ess <- emptySymState st maybeState
+                             isValidAndFailsAux c ess
+
+generateLabelsAux :: Integer -> Integer -> [String]
+generateLabelsAux n m
+  | n > m = []
+  | otherwise = (action_label ++ "minSlot"):
+                (action_label ++ "maxSlot"):
+                (action_label ++ "value"):
+                (action_label ++ "branch"):
+                generateLabelsAux (n + 1) m
+     where action_label = "action_" ++ show n ++ "_"
+
+generateLabels :: Integer -> [String]
+generateLabels = generateLabelsAux 1
+
+generateParameters :: [String] -> SymbolicT IO [(SInteger, SInteger, SInteger, SInteger)]
+generateParameters (sl:sh:v:b:t) =
+   do isl <- sInteger sl
+      ish <- sInteger sh
+      iv <- sInteger v
+      ib <- sInteger b
+      rest <- generateParameters t
+      return ((isl, ish, iv, ib):rest)
+generateParameters [] = return []
+generateParameters _ = error "Wrong number of labels generated"
+
+groupResult :: [String] -> Map String Integer -> [(Integer, Integer, Integer, Integer)]
+groupResult (sl:sh:v:b:t) mappings =
+    if ib == -1 then []
+    else (isl, ish, iv, ib):groupResult t mappings
+  where (Just isl) = M.lookup sl mappings
+        (Just ish) = M.lookup sh mappings
+        (Just iv) = M.lookup v mappings
+        (Just ib) = M.lookup b mappings
+groupResult [] _ = []
+groupResult _ _ = error "Wrong number of labels generated"
+
+caseToInput :: [Case a] -> Integer -> Integer -> Input
+caseToInput [] _ _ = error "Wrong number of cases interpreting result"
+caseToInput (Case h _:t) c v
+  | c > 1 = caseToInput t (c - 1) v
+  | c == 1 = case h of
+               Deposit accId party tok _ -> IDeposit accId party tok v
+               Choice choId _            -> IChoice choId v
+               Notify _                  -> INotify
+  | otherwise = error "Negative case number"
+
+computeAndContinue :: ([Input] -> TransactionInput) -> [Input] -> State -> Contract
+                   -> [(Integer, Integer, Integer, Integer)]
+                   -> [([TransactionInput], [TransactionWarning])]
+computeAndContinue transaction inps sta cont t =
+  case computeTransaction (transaction inps) sta cont of
+    Error TEUselessTransaction -> executeAndInterpret sta t cont
+    TransactionOutput { txOutWarnings = war
+                      , txOutState = newSta
+                      , txOutContract = newCont}
+                          ->
```

Comment:
I don't like that this params stuff is all untyped, I think it needs to be refactored however I don't want to hold up this PR so perhaps a refactoring PR could come later?

Reply:
I know, unfortunately I think part of the speed up comes from this, from just using symbolic integers for everything, instead of having symbolic lists of symbolic nested types. I tried returning that first but it really slowed things down. Maybe you know a way of wrapping everything in types that only affect the Haskell side but I am not sure how.

Reply:
Thus the use of `-1`, I would use `Maybe` in normal conditions

Reply:
ah, I meant more not using a list of strings but rather a list of records or even one record with `mapIndexed` or something.

Reply:
I especially don't like the fact that if the list is the wrong length then you `error`. This should be taken care of by the types.